### PR TITLE
update Che CRD for DWO per-user strategy storage configuration

### PR DESCRIPTION
**Pull request by AObuchow**
### What does this PR do?
This PR updates the Eclipse Che CRD regarding the workspace storage to align it with the discussion held in https://github.com/eclipse/che/issues/21405#issuecomment-1173504565

The previously named field `pvc` has been renamed to `perUserStrategyPvcConfig` to make it clear that these PVC settings only are in effect when using the "per-user" (or "common") PVC strategy. 

### Screenshot/screencast of this PR
n/a

### What issues does this PR fix or reference?
https://github.com/eclipse/che/issues/21405

### How to test this PR?
There is not much to test for this PR as it is essentially just renaming the CRD field `CheClusterDevEnvironments.Storage.Pvc` -> `CheClusterDevEnvironments.Storage.PerUserStrategyPvcConfig`.

However, this field is being used in the automated tests, so running the tests is recommended, e.g. `make test`

### PR Checklist

- [ ] [The Eclipse Contributor Agreement is valid](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-eclipse-contributor-agreement-is-valid)
- [x] [Code produced is complete](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-produced-is-complete)
- [x] [Code builds without errors](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-builds-without-errors)
- [ ] [Tests are covering the bugfix](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#tests-are-covering-the-bugfix)
- [x] [Custom resource definition file is up to date](https://github.com/eclipse-che/che-operator/blob/main/README.md#updating-custom-resource-definition-file)
- [x] [OLM bundles are up to date](https://github.com/eclipse-che/che-operator/blob/main/README.md#update-olm-bundle)
- [ ] [The repository devfile is up to date and works](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-repository-devfile-is-up-to-date-and-works)
- [x] [Sections `What issues does this PR fix or reference` and `How to test this PR` completed](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#sections-what-issues-does-this-pr-fix-or-reference-and-how-to-test-this-pr-completed)
- [ ] [Relevant user documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [ ] [Relevant contributing documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [ ] [CI/CD changes implemented, documented and communicated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#cicd-changes-implemented-documented-and-communicated)

### Reviewers

Reviewers, please comment how you tested the PR when approving it.

### CODE REVIEW

**Comment by tolusha** [comment]
Sounds good. Could you also update [1]:
```
	// +kubebuilder:default:="per-user"
	// +kubebuilder:validation:Enum=common;per-workspace;per-user
```	

[1] https://github.com/eclipse-che/che-operator/blob/885d623258450b6c942d7396c4f3c17ff05b0f2e/api/v2/checluster_types.go#L363-L364

**Comment by tolusha** [comment]
Pls, sign ECA https://accounts.eclipse.org/user/login?destination=legal/eca/validation/112175

**Comment by tolusha** [comment]
Hi. 

1. Remove `bin` directory of the `che-operator` project. It seems contains incorrect `controller-gen` tools.
```diff
 controller-gen.kubebuilder.io/version: v0.7.0
 controller-gen.kubebuilder.io/version: v0.4.1
```


2. Pls update default values here as well
https://github.com/eclipse-che/che-operator/blob/885d623258450b6c942d7396c4f3c17ff05b0f2e/api/v2/checluster_types.go#L40
https://github.com/eclipse-che/che-operator/blob/885d623258450b6c942d7396c4f3c17ff05b0f2e/api/v2/checluster_types.go#L66

3. Update dev resources

File: bundle/next/eclipse-che-preview-openshift/manifests/org.eclipse.che_checlusters.yaml [hunk 1]
```diff
@@ -14,7 +14,7 @@ apiVersion: apiextensions.k8s.io/v1
 kind: CustomResourceDefinition
 metadata:
   annotations:
-    controller-gen.kubebuilder.io/version: v0.4.1
+    controller-gen.kubebuilder.io/version: v0.7.0
```
[the remark this answers is not in the document]
Comment (AObuchow) [reply bundle/next/eclipse-che-preview-openshift/manifests/org.eclipse.che_checlusters.yaml:17]: I may be wrong, but after a quick check, it seems the makefile is [currently set up to download controller-gen@v0.7.0](https://github.com/eclipse-che/che-operator/blob/main/Makefile#L670). I can locally update it to 0.4.1?

[the remark this answers is not in the document]
Comment (tolusha) [reply bundle/next/eclipse-che-preview-openshift/manifests/org.eclipse.che_checlusters.yaml:17]: No, I've checked PR and it looks good now.

[the remark this answers is not in the document]
Comment (AObuchow) [reply bundle/next/eclipse-che-preview-openshift/manifests/org.eclipse.che_checlusters.yaml:17]: Sounds great :) Once the ECA situation is resolved, I'll squash my fix up commits for final PR approval

**Comment by AObuchow** [comment]
@tolusha Can you approve running the CI workflow please? I rebased my patch. Once the workflows succeed, I will squash the fixup commits. Thank you :)

**Comment by tolusha** [comment]
`che-server` fails with
```
Learn more:
  https://github.com/google/guice/wiki/ERROR_INJECTING_CONSTRUCTOR
Caused by: IllegalArgumentException: Unsupported PVC strategy 'per-user' configured
	at WorkspaceVolumeStrategyProvider.<init>(WorkspaceVolumeStrategyProvider.java:41)
```

**Comment by AObuchow** [comment]
@tolusha Can you please re-run tests when you have a moment, now that https://github.com/eclipse-che/che-server/pull/337 has been merged?

**Comment by tolusha** [comment]
@svor 
we have problem with starting plugin-registry container
```
Checking meta /var/www/html/v3/plugins/gattytto/dart-code/latest/che-theia-plugin.yaml
Checking meta /var/www/html/v3/plugins/lfs/vscode-emacs-friendly/latest/meta.yaml
Checking meta /var/www/html/v3/plugins/lfs/vscode-emacs-friendly/latest/che-theia-plugin.yaml
Updating relative:extension in files to http://plugin-registry.eclipse-che.svc:8080
sed: couldn't open temporary file /var/www/html/v3/plugins/felixfbecker/php-debug/latest/sedtne3gg: Permission denied
```

**Comment by svor** [comment]
Seems it's related to the latest changes in the plugin registry https://github.com/eclipse-che/che-plugin-registry/pull/1324
cc @benoitf

**Comment by tolusha** [comment]
It is ok to merge

**Review by ibuziuk — APPROVED** [review]
LGTM :+1: 
@AObuchow my understanding that the next step would be adding
```perWorkspaceStrategyPvcConfig`, right?

e.g. the final property set is expected to be the following 
```
devEnvironments:
  storage:
    pvcStrategy: <strategy>
      perUserStrategyPvcConfig:
        claimSize: <size>
        storageClass: <classname>
      perWorkspaceStrategyPvcConfig:
        claimSize: <size>
        storageClass: <classname>
```
cc: @l0rd could you please review?

File: api/v2/checluster_types.go [hunk 2]
```diff
@@ -366,15 +366,16 @@ type TrustedCerts struct {
 
 // Configuration settings related to the workspaces persistent storage.
 type WorkspaceStorage struct {
-	// PVC settings.
+	// PVC settings when using the "per-user" PVC strategy.
```
Comment (l0rd) [inline api/v2/checluster_types.go:369]: I don't think that's only for `per-user` (i.e. the `pvcStrategy` is configured in this same struct).

Comment (l0rd) [reply api/v2/checluster_types.go:369]: And I would make it clear that those settings are about the Development environments PVs.

Comment (l0rd) [reply api/v2/checluster_types.go:369]: Ok sorry my bad, the comment above is for the `perUserStrategyPvcConfig`. The diff confused me. Don't consider my comments above.

**Review by l0rd — APPROVED** [review]
LGTM

**Comment by l0rd** [comment]
> LGTM 👍 @AObuchow my understanding that the next step would be adding `perWorkspaceStrategyPvcConfig`, right?
> 
> e.g. the final property set is expected to be the following
> 
> ```
> devEnvironments:
>   storage:
>     pvcStrategy: <strategy>
>       perUserStrategyPvcConfig:
>         claimSize: <size>
>         storageClass: <classname>
>       perWorkspaceStrategyPvcConfig:
>         claimSize: <size>
>         storageClass: <classname>
> ```
> 
> cc: @l0rd could you please review?

@ibuziuk that's fine to add the `perWorkspaceStrategyPvcConfig` because I am sure that it will be requested. But we should not invest on the `per-workspace` strategy beyond that. The `workspace-strategy` is a workaround to let users run multiple workspaces at the same time until we have good support for async storage. My understanding is that with async storage we won't need `per-workspace` strategy anymore.

**Comment by AObuchow** [comment]
> LGTM +1 @AObuchow my understanding that the next step would be adding `perWorkspaceStrategyPvcConfig`, right?
> 
> e.g. the final property set is expected to be the following
> 
> ```
> devEnvironments:
>   storage:
>     pvcStrategy: <strategy>
>       perUserStrategyPvcConfig:
>         claimSize: <size>
>         storageClass: <classname>
>       perWorkspaceStrategyPvcConfig:
>         claimSize: <size>
>         storageClass: <classname>
> ```
> 
> cc: @l0rd could you please review?

Sounds good @ibuziuk, I will submit a followup PR to add the equivalent configuration for the perWorkspace PVC strategy

**Pull request merged**